nexus/src/internal_api/http_entrypoints.rs

```diff
@@ -280,3 +285,76 @@ async fn cpapi_metrics_collect(
         .instrument_dropshot_handler(&request_context, handler)
         .await
 }
+
+#[derive(Deserialize, JsonSchema)]
+struct AllPath {
+    path: String,
+}
+
+/// Endpoint used by Sled Agents to download cached artifacts.
+#[endpoint {
+    method = GET,
+    path = "/artifacts/{path}",
+}]
+async fn cpapi_artifact_download(
+    request_context: Arc<RequestContext<Arc<ServerContext>>>,
+    path: Path<AllPath>,
+) -> Result<Response<Body>, HttpError> {
+    let context = request_context.context();
+    let nexus = &context.nexus;
+    let mut entry = PathBuf::from(BASE_ARTIFACT_DIR);
+
+    // TODO: Most of the below code is ready to accept a multi-component path,
+    // such as in:
+    // https://github.com/oxidecomputer/dropshot/blob/78be3deda556a9339ea09f3a9961fd91389f8757/dropshot/examples/file_server.rs#L86-L89
+    //
+    // However, openapi does *not* like that currently, so we limit the endpoint
+    // to only accepting single-component paths.
+    let path = vec![path.into_inner().path];
+
+    for component in &path {
+        // Dropshot should not provide "." and ".." components.
+        assert_ne!(component, ".");
+        assert_ne!(component, "..");
+        entry.push(component);
+
+        if entry.exists() {
+            // We explicitly prohibit consumers from following symlinks to prevent
+            // showing data outside of the intended directory.
+            let m = entry.symlink_metadata().map_err(|e| {
+                HttpError::for_bad_request(
+                    None,
+                    format!("Failed to query file metadata: {}", e),
+                )
+            })?;
+            if m.file_type().is_symlink() {
+                return Err(HttpError::for_bad_request(
+                    None,
+                    "Cannot traverse symlinks".to_string(),
+                ));
+            }
+        }
+    }
+
+    // Note - at this point, "entry" may or may not actually exist.
+    // We try to avoid creating intermediate artifacts until we know there
+    // is something "real" to download, as this would let malformed paths
+    // create defunct intermediate directories.
+    if entry.is_dir() {
+        return Err(HttpError::for_bad_request(
+            None,
+            "Directory download not supported".to_string(),
+        ));
+    }
+    let body = nexus.download_artifact(&entry).await?;
```

Comment:
Isn't there a TOCTOU race here, if the path has changed since we checked it?

I think the traditional approach is to use open(2) and then fstat(2).  I haven't dug into it but it looks like you can do this in Rust with `File::open` and then `File::metadata`.

Reply:
... this may need an update: https://github.com/oxidecomputer/dropshot/blob/main/dropshot/examples/file_server.rs#L132

Reply:
So I'm looking at the file server - I think that whole example may need a re-work, in addition to this "final entry" check. We also check that the intermediate paths are not symlinks, and I believe this suffers from the same TOCTTOU issue - they could be replaced by symlinks, since we're re-checking the whole path again.

In C, I would normally use `openat` to resolve this issue, but I don't see this function exposed from the Rust standard library (though I suppose I could use libc directly, or one of the wrapper crates).